### ISSUE TITLE
controls_set already set where can I add a control element?

### ISSUE THREAD

**Issue by http-teapot** [spolu/breach_core#158]
I am trying to create an autocomplete feature for mod_strip and I've been digging around to see if it was possible but it seems that controls can only be added within 4 positions therefore only a total of 4 controls can be added (I might be wrong).

Can I add a element to Breach and position it anywhere I want?
Can I extend mod_strip to do what I want or should I fork mod_strip?

I was able to listen to mod_strip's events but couldn't figure out how to extend mod_strip control elements.

Here is my repo (which looks a lot like mod_strip, might do some cleaning later, or maybe find a smart way to consolidate how visual elements are rendered)
https://github.com/http-teapot/mod_duckduckgo/tree/master

**Comment by brayniverse** [comment]
@http-teapot It would appear that you will need to fork mod_strip in order to extend it's features to include autocomplete functionalities.

From what I gathered you are correct and there are only 4 places a control can be positioned; top, bottom, left and right - and I see no indication whether there are plans to add more yet. Although aside from mod_strip I haven't seen anybody make use of the core_controls module yet and I think when that starts happening we will have more clarity on how it will progress.

Also each module broadcasts events so that other modules can react accordingly but as far as I can tell there is no way of actually extending the UI of a control. Which would be nice to do if we will only have 4 sections to occupy.

**Comment by spolu** [comment]
Hi! There is a floating control expose by the exo_browser that is not yet exposed in breach_core. It's fairly easy to add and would probably help with your autocomplete module.

I'll try to add it next week so that you can make progress on that.

That being said, the goal for breach is to move to the use of a <webview> tag that would enable much more flexibility in how the controls and tabs are disposed together. 

Basically modules would be able to inject code into the main document to position tabs <webview>a and inject iframes as controllers. The problem of extending a control stay entirely open in that new setup and any brilliant idea related to that is definitely welcome!! :-)

**Comment by http-teapot** [comment]
Hello!

I actually was digging into exo_browser to find a hint of this feature but kind of got stuck on the 4 position constant system. I might spend a bit more time figuring it out today. (Edit: After a good night sleep, I found the method `show_floating` defined under `set_control` so I might play with it)

About modules injecting code in the document, are you talking about template inheritance? That would be definitely one way to go.
Otherwise, before being able to extend other modules or even talking about template inheritance, it would be great to consolidate how UI elements are rendered (for example, spawning a server instance) and maybe surface a mini-API for each rendered elements (for instance, to inject assets).
I can probably help on that.

I look forward to these changes! :)

**Comment by spolu** [comment]
Hey @http-teapot thanks fo the details here.

Could you describe with a little bit more details what kind of API you're thinking of? I'm curious!

**Comment by spolu** [comment]
I've looked into exposing the `floating` control but it's really not ready for prime time yet. This will be entirely solved in the next release of Breach as explained above.

**Comment by spolu** [comment]
Closing for now as this was discussed.
Will be covered by the roadmap

**Comment by http-teapot** [comment]
I felt a bit overwhelmed by how the backend is structured and how it can interacts with the front-end: anything is possible.
It would be great to come up with a default architecture (or an advised one) that would take consolidate the modules.

In terms of UI frameworks, you could look at [Flight.js](https://github.com/flightjs/flight) or [React.js](https://github.com/facebook/react) (includes rendering). Flight.js is event driven while React.js is a powerful component renderer and they both can unify front-end layer.

I don't have any suggestions for a back-end framework; but consolidating the following would be great:
- view rendering
- express initializing
- socket.io initializing and handling (making it seamless between both back-end/front-end)
- component handling through `common._.` syntax

I really like how breach_core interacts with breach modules, would love if the module's back-end and front-end would behave the same way.

**Comment by spolu** [comment]
@http-teapot module consolidation is definitely a very complicated subject and one that will require a lot of work to get right. The upcoming architecture will allow for a great flexibility at the ExoBrowser level. 

We need to decide what API is exposed to Breach module, how to suggest collaboration between modules and incentivize it, and at the same time let some doors open for modules to create entirely specific solutions. As an example using famo.us or openGL which would not work well with anything else!

**Comment by http-teapot** [comment]
famo.us would be a pretty cool framework for visual interaction and they are planning to support OpenGL.

I just tried creating a atom.io module and it's pretty interesting. To a certain extent, Coffeescript adds the ability to easily create classes and extend them. Not super convinced about the syntax itself but it's pretty cool.

**Comment by spolu** [comment]
They also constrain you to some type of views right?

**Comment by http-teapot** [comment]
I didn't dig too much on that end but by default it seems to generate HTML through a View class. It's probably easy to plug in a template engine of some kind.

**Comment by spolu** [comment]
But the positioning of the view is constrainted I presume?

**Comment by http-teapot** [comment]
Actually, there is a WorkspaceView which holds all other Views: some basic Views can be used or a custom one can be made. I didn't dig too much into creating a visual module but from I read it's pretty flexible.
https://atom.io/docs/v0.120.0/advanced/view-system

The theming system is really interesting too: https://atom.io/docs/v0.120.0/creating-a-theme
Atom.io provides a set of variables in order to keep colors consistent across modules.

**Comment by spolu** [comment]
This is really interesting. I think it's time to start working on a spec for the new breach API. At least get some ideas and concept written down so that we can discuss them. We should probably start an issue for that...

Anybody wants to take a stab at it?
